WordPress/src/main/java/org/wordpress/android/ui/stockmedia/StockMediaPickerActivity.java

```diff
@@ -0,0 +1,590 @@
+                if (isEmpty) {
+                    String message = getString(R.string.stock_media_picker_initial_empty_text);
+                    String subMessage = getString(R.string.stock_media_picker_initial_empty_subtext);
+                    String link = "<a href='https://pexels.com/'>Pexels</a>";
+                    String html = message
+                                  + "<br /><br />"
+                                  + "<small>" + String.format(subMessage, link) + "</small>";
+                    txtEmpty.setMovementMethod(WPLinkMovementMethod.getInstance());
+                    txtEmpty.setText(Html.fromHtml(html));
+                } else {
+                    txtEmpty.setText(R.string.stock_media_picker_empty_results);
+                }
+            }
+        }
+    }
+
+    private void showProgress(boolean show) {
+        if (!isFinishing()) {
+            findViewById(R.id.progress).setVisibility(show ? View.VISIBLE : View.GONE);
+        }
+    }
+
+    private void submitSearch(@Nullable final String query, boolean delayed) {
+        mSearchQuery = query;
+
+        if (query == null || query.length() < MIN_SEARCH_QUERY_SIZE) {
+            mAdapter.clear();
+            showEmptyView(true);
+            return;
+        }
+
+        if (delayed) {
+            mHandler.postDelayed(new Runnable() {
+                @Override
+                public void run() {
+                    if (StringUtils.equals(query, mSearchQuery)) {
+                        submitSearch(query, false);
+                    }
+                }
+            }, 500);
+        } else {
+            fetchStockMedia(query, 1);
+        }
+    }
+
+    private void fetchStockMedia(@Nullable String searchQuery, int page) {
+        if (!NetworkUtils.checkConnection(this)) return;
+
+        if (TextUtils.isEmpty(searchQuery)) {
+            mAdapter.clear();
+            return;
+        }
+
+        if (page == 1) {
+            mAdapter.clear();
+        }
+
+        showProgress(true);
+
+        mIsFetching = true;
+
+        mSearchQuery = searchQuery;
+        AppLog.d(AppLog.T.MEDIA, "Fetching stock media page " + page);
+
+        StockMediaStore.FetchStockMediaListPayload payload =
+                new StockMediaStore.FetchStockMediaListPayload(searchQuery, page);
+        mDispatcher.dispatch(StockMediaActionBuilder.newFetchStockMediaAction(payload));
+    }
+
+    @SuppressWarnings("unused")
+    @Subscribe(threadMode = ThreadMode.MAIN)
+    public void onStockMediaListFetched(StockMediaStore.OnStockMediaListFetched event) {
+        mIsFetching = false;
+        showProgress(false);
+
+        if (event.isError()) {
+            AppLog.e(AppLog.T.MEDIA, "An error occurred while searching stock media");
+            mCanLoadMore = false;
+            return;
+        }
+
+        // make sure these results are for the same query
+        if (mSearchQuery == null || !mSearchQuery.equals(event.searchTerm)) {
```

Comment:
I think this should happen before we do anything else. Let's say that user is a slow writer and their connection is slow. The user writes `rando` and starts a fetch and keeps writing after to complete to `random` before the initial fetch results are received. In this case, we are actually interested in `random` but we receive the result for `rando` and we already mark it as `mIsFetching = false` and handle errors related to it.

Hope that makes sense!